### PR TITLE
ISSUE-2037: add expression transform optimizer

### LINKED ISSUE #2037: [improvement] add truth transform for expression
**Summary**

Apply NOT transformation to the expression and return a new one.

Transform expression item using next rules:
```
       a AND b AND -> NOT(a) OR NOT(b) OR ...
       a OR b OR -> NOT(a) AND NOT(b) AND ...
       NOT(a) -> a
       a = b -> a != b
       a != b -> a = b
       a < b -> a >= b
       a >= b -> a < b
       a > b -> a <= b
       a <= b -> a > b
       a LIKE b -> a NOT LIKE b
       a NOT LIKE b -> a LIKE b 
       IS NULL(a) -> IS NOT NULL(a)
       IS NOT NULL(a) -> IS NULL(a)
```

### PR DESCRIPTION
**Pull request by zhyass**
I hereby agree to the terms of the CLA available at: https://databend.rs/policies/cla/

## Summary

Add expression transform optimizer, support NOT transformer

## Changelog

- New Feature
- Improvement

## Related Issues

Fixes #2037 

## Test Plan

Unit Tests

### CODE REVIEW

File: query/src/optimizers/utils.rs [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+                let mut v = self;
+                v.required_columns.insert(c.clone());
+                Ok(Recursion::Continue(v))
+            }
+            _ => Ok(Recursion::Continue(self)),
+        }
+    }
+}
+
+pub fn create_scalar_function(op: &str, args: Expressions) -> Expression {
```
Comment (zhang2014) [inline query/src/optimizers/utils.rs:47]: Maybe move `create_scalar_function` , `create_unary_expression`,  `create_binary_expression` to Expression is better?

File: query/src/optimizers/optimizer_expression_transform.rs [hunk 2]
```diff
@@ -0,0 +1,200 @@
...
+        }
+
+        let new_op = INVERSE_OPERATOR.get(op);
+        match new_op {
+            Some(v) => f(v, args),
+            None => create_unary_expression("NOT", vec![origin.clone()]),
+        }
+    }
+
+    fn truth_transformer(origin: &Expression, is_negated: bool) -> Expression {
```
Comment (zhang2014) [inline query/src/optimizers/optimizer_expression_transform.rs:79]: Maybe we can add `convert_to_negative` to Function trait.
```
trait Function {
    ...
    pub fn convert_to_negative(&self) -> Option<Self>;
    ...
}
```

File: query/src/optimizers/utils.rs (hunk 1 above)
Comment (zhyass) [reply query/src/optimizers/utils.rs]: Good suggestion.

File: query/src/optimizers/optimizer_expression_transform.rs [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+lazy_static! {
+    static ref INVERSE_OPERATOR: HashMap<&'static str, &'static str> = {
+        let mut map = HashMap::new();
+        map.insert("=", "<>");
+        map.insert("<>", "=");
+        map.insert("<", ">=");
+        map.insert("<=", ">");
+        map.insert(">", "<=");
+        map.insert(">=", "<");
+        map.insert("LIKE", "NOT LIKE");
```
Comment (sundy-li) [inline query/src/optimizers/optimizer_expression_transform.rs:47]: we should consider case insensitive

Comment (zhyass) [reply query/src/optimizers/optimizer_expression_transform.rs:47]: https://github.com/datafuselabs/databend/pull/2045

**Pull request merged**